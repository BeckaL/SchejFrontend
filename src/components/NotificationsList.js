import React, { Component } from 'react'
import axios from 'axios'

export default class NotificationsList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      requests: [],
      emergencies: []
    }
  }

  componentWillMount() {
    axios.get(
      `${process.env.REACT_APP_API_URL}/requestsbyuser/${localStorage['id']}`
    )
    .then(response => {
      this.setState({ requests: response.data })
    })

    setTimeout(() => {
      axios.get(
        `${process.env.REACT_APP_API_URL}/emergency_requests?user_id=${localStorage['id']}`
      )
      .then(response => {
        console.log(response)
        this.setState({ emergencies: response.data })
      })
    }, 20)
  }

  handleEmergencyApprove = (event) => {
    const index = event.target.attributes.index.value
    let emergencies = [...this.state.emergencies]
    let emergency = {...emergencies[index]}
    emergency.approved = true
    emergencies[index] = emergency
    this.setState({emergencies})

    axios.patch(
      `${process.env.REACT_APP_API_URL}/emergencyshift/${this.state.emergencies[index].shiftId}`, {
        respondent_id: localStorage['id']
      }
    )

    this.deleteEmergencyRequest(this.state.emergencies[index].id)
  }

  handleApprove = (event) => {
    const index = event.target.attributes.index.value
    let requests = [...this.state.requests]
    let request = {...requests[index]}
    request.approved = true
    requests[index] = request
    this.setState({requests})

    axios.patch(
      `${process.env.REACT_APP_API_URL}/shifts/${this.state.requests[index].requesterShift.id}?other_id=${this.state.requests[index].respondentShift.id}`
    )
    .then(response => {
      console.log(response)
    })

    this.deleteRequest(this.state.requests[index].id)
  }

  handleDecline = (event) => {
    const index = event.target.attributes.index.value
    let requests = [...this.state.requests]
    let request = {...requests[index]}
    request.declined = true
    requests[index] = request
    this.setState({requests})

    this.deleteRequest(this.state.requests[index].id)
  }

  deleteEmergencyRequest = (emergencyId) => {
    axios.delete(
      `${process.env.REACT_APP_API_URL}/emergency_requests/${emergencyId}`
    )
  }

  deleteRequest = (requestId) => {
    axios.delete(
      `${process.env.REACT_APP_API_URL}/requests/${requestId}`
    )
    .then(response => {
      console.log(response)
    })
    .catch(error => {
      console.log(error)
    })
  }

  formatDate = (time) => {
    return new Date(parseInt(time)).toLocaleString("en-GB", { day: '2-digit', month: 'short', hour: '2-digit', minute: '2-digit'})
  }

  formatEmergencyRequestContent = () => {
    return this.state.emergencies.map((emergency, index) => {
      if (emergency.approved) {
        return <div id="emergency-box" key={index}><span>Thanks! You have now taken on this shift. You are a hero!</span></div>
      } else {
      return (
        <div id="emergency-box" key={index}>
          <span id="emergency-message">Comment: {this.state.emergencies[index].comment}</span><br />
          <span>{this.state.emergencies[index].name} has an emergency and cannot make their shift on </span>
          <span>{this.formatDate(this.state.emergencies[index].start)} until </span>
          <span>{this.formatDate(this.state.emergencies[index].end)}</span><br />
          <button id="approve-emergency-button" className="custom-button" index={index} onClick={this.handleEmergencyApprove}>Approve</button><br /><br />
        </div>
      )
    }
    })
  }

  formatRequestContent = () => {
    return this.state.requests.map((request, index) => {
      if (request.approved) {
        return <div id="approve-notification-box" key={index}><span>Thanks! Your shifts have been swapped.</span></div>
      } else if (request.declined){
        return <div id="decline-notification-box" key={index}><span>You have declined to swap.</span></div>
      } else {
        return (
          <div id="notification-box" key={index}>
            <span id="notification-message"><b>{this.state.requests[index].requesterShift.name}</b> wants to swap shifts with you.</span><br />
            <span id="requester-shift-details">Their shift: {this.formatDate(this.state.requests[index].requesterShift.start)} - {this.formatDate(this.state.requests[index].requesterShift.end)}</span><br />
            <span id="respondent-shift-details">Your shift: {this.formatDate(this.state.requests[index].respondentShift.start)} - {this.formatDate(this.state.requests[index].respondentShift.end)}</span><br />
            <button id="approve-swap-button" className="custom-button" index={index} onClick={this.handleApprove}>Approve</button>
            <button id="decline-swap-button" className="custom-button" index={index} onClick={this.handleDecline}>Decline</button><br /><br />
          </div>
        )
      }
    })
  }

  render() {
    return(
      <div>
        <h3 className='popup-title'>Notifications</h3>
        <div id='notifications-list'>
          {this.formatRequestContent()}
          <span><b>Emergency Requests:</b></span>
          <br />
          {this.formatEmergencyRequestContent()}
        </div>
      </div>
  )}
}
