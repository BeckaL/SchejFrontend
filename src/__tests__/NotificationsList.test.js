import React from 'react'
import { create } from "react-test-renderer"
import { mount, shallow } from 'enzyme'
import axios from 'axios'
import NotificationsList from '../components/NotificationsList'

jest.mock('axios')

describe('Notifications List', () => {
  var wrapper
  var initialResponse

  beforeAll(() => {
    initialResponse = { data: [] }
    axios.get.mockResolvedValue(initialResponse)
    wrapper = shallow(<NotificationsList history = {[]} />)
  })

  test("it matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  test("it renders a notifications list", () => {
    const divs = wrapper.find('div')
    expect(divs.length).toEqual(2)
  })

  test("it renders a notification list with notifications", () => {
    const newResponse = { data: [1,2,3,4,5] }
    axios.get.mockResolvedValue(newResponse)
    const divs = wrapper.find('div')
    expect(divs.length).toEqual(7)
  })
})
