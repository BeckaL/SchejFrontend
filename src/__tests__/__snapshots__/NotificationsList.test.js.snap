// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications List it matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationsList
    history={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="popup-title"
        >
          Notifications
        </h3>,
        <div
          id="notifications-list"
        >
          <div
            id="notification-box"
          >
            <span
              id="notification-message"
            >
              <b>
                TestName2
              </b>
               wants to swap shifts with you.
            </span>
            <br />
            <span
              id="requester-shift-details"
            >
              Their shift: 
              18 Dec, 16:49
               - 
              18 Dec, 16:49
            </span>
            <br />
            <span
              id="respondent-shift-details"
            >
              Your shift: 
              18 Dec, 16:49
               - 
              18 Dec, 16:49
            </span>
            <br />
            <button
              className="custom-button"
              id="decline-swap-button"
              index={0}
              onClick={[Function]}
            >
              Decline
            </button>
            <button
              className="custom-button"
              id="approve-swap-button"
              index={0}
              onClick={[Function]}
            >
              Approve
            </button>
            <br />
            <br />
          </div>
          <span>
            <b>
              Emergency Requests:
            </b>
          </span>
          <br />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Notifications",
          "className": "popup-title",
        },
        "ref": null,
        "rendered": "Notifications",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <div
                id="notification-box"
              >
                <span
                  id="notification-message"
                >
                  <b>
                    TestName2
                  </b>
                   wants to swap shifts with you.
                </span>
                <br />
                <span
                  id="requester-shift-details"
                >
                  Their shift: 
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </span>
                <br />
                <span
                  id="respondent-shift-details"
                >
                  Your shift: 
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </span>
                <br />
                <button
                  className="custom-button"
                  id="decline-swap-button"
                  index={0}
                  onClick={[Function]}
                >
                  Decline
                </button>
                <button
                  className="custom-button"
                  id="approve-swap-button"
                  index={0}
                  onClick={[Function]}
                >
                  Approve
                </button>
                <br />
                <br />
              </div>,
            ],
            <span>
              <b>
                Emergency Requests:
              </b>
            </span>,
            <br />,
            Array [],
          ],
          "id": "notifications-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  id="notification-message"
                >
                  <b>
                    TestName2
                  </b>
                   wants to swap shifts with you.
                </span>,
                <br />,
                <span
                  id="requester-shift-details"
                >
                  Their shift: 
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </span>,
                <br />,
                <span
                  id="respondent-shift-details"
                >
                  Your shift: 
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </span>,
                <br />,
                <button
                  className="custom-button"
                  id="decline-swap-button"
                  index={0}
                  onClick={[Function]}
                >
                  Decline
                </button>,
                <button
                  className="custom-button"
                  id="approve-swap-button"
                  index={0}
                  onClick={[Function]}
                >
                  Approve
                </button>,
                <br />,
                <br />,
              ],
              "id": "notification-box",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <b>
                      TestName2
                    </b>,
                    " wants to swap shifts with you.",
                  ],
                  "id": "notification-message",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "TestName2",
                    },
                    "ref": null,
                    "rendered": "TestName2",
                    "type": "b",
                  },
                  " wants to swap shifts with you.",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Their shift: ",
                    "18 Dec, 16:49",
                    " - ",
                    "18 Dec, 16:49",
                  ],
                  "id": "requester-shift-details",
                },
                "ref": null,
                "rendered": Array [
                  "Their shift: ",
                  "18 Dec, 16:49",
                  " - ",
                  "18 Dec, 16:49",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Your shift: ",
                    "18 Dec, 16:49",
                    " - ",
                    "18 Dec, 16:49",
                  ],
                  "id": "respondent-shift-details",
                },
                "ref": null,
                "rendered": Array [
                  "Your shift: ",
                  "18 Dec, 16:49",
                  " - ",
                  "18 Dec, 16:49",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Decline",
                  "className": "custom-button",
                  "id": "decline-swap-button",
                  "index": 0,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Decline",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Approve",
                  "className": "custom-button",
                  "id": "approve-swap-button",
                  "index": 0,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Approve",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <b>
                Emergency Requests:
              </b>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Emergency Requests:",
              },
              "ref": null,
              "rendered": "Emergency Requests:",
              "type": "b",
            },
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="popup-title"
          >
            Notifications
          </h3>,
          <div
            id="notifications-list"
          >
            <div
              id="notification-box"
            >
              <span
                id="notification-message"
              >
                <b>
                  TestName2
                </b>
                 wants to swap shifts with you.
              </span>
              <br />
              <span
                id="requester-shift-details"
              >
                Their shift: 
                18 Dec, 16:49
                 - 
                18 Dec, 16:49
              </span>
              <br />
              <span
                id="respondent-shift-details"
              >
                Your shift: 
                18 Dec, 16:49
                 - 
                18 Dec, 16:49
              </span>
              <br />
              <button
                className="custom-button"
                id="decline-swap-button"
                index={0}
                onClick={[Function]}
              >
                Decline
              </button>
              <button
                className="custom-button"
                id="approve-swap-button"
                index={0}
                onClick={[Function]}
              >
                Approve
              </button>
              <br />
              <br />
            </div>
            <span>
              <b>
                Emergency Requests:
              </b>
            </span>
            <br />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Notifications",
            "className": "popup-title",
          },
          "ref": null,
          "rendered": "Notifications",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <div
                  id="notification-box"
                >
                  <span
                    id="notification-message"
                  >
                    <b>
                      TestName2
                    </b>
                     wants to swap shifts with you.
                  </span>
                  <br />
                  <span
                    id="requester-shift-details"
                  >
                    Their shift: 
                    18 Dec, 16:49
                     - 
                    18 Dec, 16:49
                  </span>
                  <br />
                  <span
                    id="respondent-shift-details"
                  >
                    Your shift: 
                    18 Dec, 16:49
                     - 
                    18 Dec, 16:49
                  </span>
                  <br />
                  <button
                    className="custom-button"
                    id="decline-swap-button"
                    index={0}
                    onClick={[Function]}
                  >
                    Decline
                  </button>
                  <button
                    className="custom-button"
                    id="approve-swap-button"
                    index={0}
                    onClick={[Function]}
                  >
                    Approve
                  </button>
                  <br />
                  <br />
                </div>,
              ],
              <span>
                <b>
                  Emergency Requests:
                </b>
              </span>,
              <br />,
              Array [],
            ],
            "id": "notifications-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    id="notification-message"
                  >
                    <b>
                      TestName2
                    </b>
                     wants to swap shifts with you.
                  </span>,
                  <br />,
                  <span
                    id="requester-shift-details"
                  >
                    Their shift: 
                    18 Dec, 16:49
                     - 
                    18 Dec, 16:49
                  </span>,
                  <br />,
                  <span
                    id="respondent-shift-details"
                  >
                    Your shift: 
                    18 Dec, 16:49
                     - 
                    18 Dec, 16:49
                  </span>,
                  <br />,
                  <button
                    className="custom-button"
                    id="decline-swap-button"
                    index={0}
                    onClick={[Function]}
                  >
                    Decline
                  </button>,
                  <button
                    className="custom-button"
                    id="approve-swap-button"
                    index={0}
                    onClick={[Function]}
                  >
                    Approve
                  </button>,
                  <br />,
                  <br />,
                ],
                "id": "notification-box",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <b>
                        TestName2
                      </b>,
                      " wants to swap shifts with you.",
                    ],
                    "id": "notification-message",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "TestName2",
                      },
                      "ref": null,
                      "rendered": "TestName2",
                      "type": "b",
                    },
                    " wants to swap shifts with you.",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Their shift: ",
                      "18 Dec, 16:49",
                      " - ",
                      "18 Dec, 16:49",
                    ],
                    "id": "requester-shift-details",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Their shift: ",
                    "18 Dec, 16:49",
                    " - ",
                    "18 Dec, 16:49",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Your shift: ",
                      "18 Dec, 16:49",
                      " - ",
                      "18 Dec, 16:49",
                    ],
                    "id": "respondent-shift-details",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Your shift: ",
                    "18 Dec, 16:49",
                    " - ",
                    "18 Dec, 16:49",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Decline",
                    "className": "custom-button",
                    "id": "decline-swap-button",
                    "index": 0,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Decline",
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Approve",
                    "className": "custom-button",
                    "id": "approve-swap-button",
                    "index": 0,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Approve",
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <b>
                  Emergency Requests:
                </b>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Emergency Requests:",
                },
                "ref": null,
                "rendered": "Emergency Requests:",
                "type": "b",
              },
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
